v3/src/execution/__tests__/execution.test.ts

```diff
@@ -0,0 +1,366 @@
+      query Basic {
+        invalidField
+      }
+    `
+    }
+  ],
+  [
+    "with invalid selection sets",
+    {
+      query: gql`
+      query Nested {
+        people {
+          name
+          invalidField
+        }
+      }
+    `
+    }
+  ],
+  [
+    "missing directive",
+    {
+      query: gql`
+      query MissingDirective {
+        people @invalid {
+          name
+        }
+      }
+      `
+    }
+  ],
+  [
+    "with multiple errors",
+    {
+      query: gql`
+      query FirstDocument {
+        people {
+          invalidOne
+        }
+        invalidTwo
+      }
+    `
+    }
+  ]
+];
+
+const nonExecutableQueries: JestInput<{ query: string }> = [
+  ['error field throws', {
+    query: gql`
+      query {
+        error
+      }
+    `
+  }],
+  ['throws multiple errors', {
+    query: gql`
+      query {
+        error
+        people {
+          nestedError
+        }
+      }
+    `
+  }]
+]
+
+const serverTeam = [
+  { id: 1, name: "James Baxley" },
+  { id: 2, name: "Ashi Krishnan" },
+  { id: 3, name: "Jesse Rosenberger" },
+  { id: 4, name: "Trevor Scheer" }
+];
+
+const schema = makeExecutableSchema({
+  typeDefs: gql`
+    type Query {
+      people: [Person]
+      error: String
+    }
+
+    type Person {
+      id: ID
+      name: String
+      friends: [Person]
+      nestedError: String
+    }
+  `,
+  resolvers: {
+    Query: {
+      people() {
+        return serverTeam;
+      },
+      error() {
+        throw new GraphQLError("Error while resolving `error` field");
+      }
+    },
+    Person: {
+      async friends(parentValue) {
+        return serverTeam.filter(({ id }) => id !== parentValue.id);
+      },
+      nestedError() {
+        throw new GraphQLError("Error while resolving `nestedError` field")
+      }
+    }
+  }
+});
+
+describe("parseGraphqlRequest", () => {
+  it.each(validQueries)("Parses valid queries - %s", (_, { query }) => {
+    const parsed = parseGraphqlRequest({ query });
+    expect(parsed).toMatchSnapshot();
+  });
+
+  it.each(unparseableQueries)("Throws for invalid queries - %s", (_, query) => {
+    expect(() => parseGraphqlRequest({ query })).toThrowErrorMatchingSnapshot();
+  });
+});
+
+describe("validateGraphqlRequest", () => {
+  it.each(validQueries)(
+    "valid queries against schema: %s",
+    (_, { query, operationName }) => {
+      const document = parseGraphqlRequest({ query });
+      expect(() => validateGraphqlRequest({
+        schema,
+        document,
+        operationName
+      })).not.toThrow();
+    }
+  );
+
+  it.each(invalidQueries)(
+    "invalid queries against schema: %s",
+    (_, { query, operationName }) => {
+      const document = parseGraphqlRequest({ query });
+
+      try {
+        validateGraphqlRequest({
+          schema,
+          document,
+          operationName
+        });
+      } catch (e) {
+        // Jest's `expect().toThrowErrorMatchingSnapshot()` doesn't work for throwing an array of errors
```

Comment:
On this note, should we be throwing an _array_ of `GraphQLError`s? This is a bit smelly to me just because it's unusual, though it's not problematic AFAICT.

Reply:
We're now catching and returning errors in all cases rather than throwing anything, which is what `ApolloServer` currently does.